### ISSUE TITLE
`#![deny(unsafe_code)]` does not detect unsafe code generated by macro from a different crate.

### ISSUE THREAD

**Issue by ZetaNumbers** [rust-lang/rust#88987]
I tried this code:

```rust
// macro_impl/src/lib.rs
#![deny(unsafe_code)]

#[macro_export]
macro_rules! unsafe_macro {
    () => {
        unsafe { *core::ptr::null_mut::<i32>() }
    };
}

// src/main.rs
#![deny(unsafe_code)]

fn main() {
    println!("{}", macro_impl::unsafe_macro!());
}
```

I expected to see this happen: error at the `unsafe_macro` invocation 

Instead, this happened: it compiles

### Meta

`rustc --version --verbose`:
```
rustc 1.57.0-nightly (c3c0f80d6 2021-09-14)
binary: rustc
commit-hash: c3c0f80d6081092faff801542dd82f0e2420152b
commit-date: 2021-09-14
host: x86_64-unknown-linux-gnu
release: 1.57.0-nightly
LLVM version: 13.0.0
```

<details>
<summary>cargo expand</summary>

```rust
#![feature(prelude_import)]
#![deny(unsafe_code)]
#[prelude_import]
use std::prelude::rust_2021::*;
#[macro_use]
extern crate std;
fn main() {
    {
        ::std::io::_print(
            match match () {
                () => [],
            } {
                ref args => unsafe { ::core::fmt::Arguments::new_v1(&["Hello\n"], args) },
            },
        );
    };
    {
        ::std::io::_print(
            match match (&unsafe { *core::ptr::null_mut::<i32>() },) {
                (arg0,) => [::core::fmt::ArgumentV1::new(
                    arg0,
                    ::core::fmt::Display::fmt,
                )],
            } {
                ref args => unsafe { ::core::fmt::Arguments::new_v1(&["", "\n"], args) },
            },
        );
    };
    {
        ::std::io::_print(
            match match () {
                () => [],
            } {
                ref args => unsafe { ::core::fmt::Arguments::new_v1(&["Hello\n"], args) },
            },
        );
    };
}
```

</details>

<details><summary>Backtrace</summary>
<p>

```
<backtrace>
```

</p>
</details>

**Comment by ZetaNumbers** [comment]
After closer inspection of `println!`'s generated code, i now understand why this is a necessity.

**Comment by PatchMixolydic** [comment]
I presume that this is for the same reason `#![deny(unsafe_code)]` doesn't disallow the use of `Vec::push`: public macros that use `unsafe` internally are meant to be safe abstractions for unsafe code, and should be memory safe regardless of the input given to them. To export a macro that could cause UB in safe Rust (ex. having a public macro `to_bool!($x:ident)` that expands to `unsafe { std::mem::transmute::<u8, bool>($x) }`) is an incorrect usage of `unsafe` code and is thus the fault of the macro author, just as a function that can cause UB when called from safe Rust is incorrect and is the fault of the function author.
